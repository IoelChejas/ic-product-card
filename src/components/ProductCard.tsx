import { useProduct } from '../hooks/useProduct'
import { OnChangeProductCountArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/product.interface'

import styles from '../styles/styles.module.css'
import React, { createContext, ReactElement } from 'react';

import { ProductCardInitialValues } from '../interfaces/product.interface'

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext

export interface ProductCardProps {
    product: Product;
    children?: ReactElement | ReactElement[] | ((args: ProductCardHandlers) => JSX.Element);
    className?: string;
    style?: React.CSSProperties;
    onChange?: (args: OnChangeProductCountArgs) => void;
    value?: number;
    initialValues?: ProductCardInitialValues
}

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: ProductCardProps) => {
    const { counter, increaseBy, reset, maxCount, isMaxCountReached } = useProduct({ product, onChange, value, initialValues })

    return (
        <Provider value={{
            counter,
            product,
            increaseBy,
            maxCount
        }}>
            <div
                className={`${styles.productCard} ${className}`}
                style={style}
            >
                {(children && typeof children === 'function') ?
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount: initialValues?.maxCount,
                        product,
                        increaseBy,
                        reset
                    })
                    : children}
            </div>
        </Provider>
    )
}

