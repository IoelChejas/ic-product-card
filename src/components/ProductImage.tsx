import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import NoImage from '../assets/no-image.jpg'
import styles from '../styles/styles.module.css'

export interface ProductImageProps {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductImage = ({ img = '', className, style }: ProductImageProps) => {

    const { product } = useContext(ProductContext)

    let imgToShow: string;
    if (img) {
        imgToShow = img;
    } else if (product.img) {
        imgToShow = product.img
    } else {
        imgToShow = NoImage
    }

    return (
        <img
            className={`${styles.productImg} ${className}`}
            src={imgToShow}
            alt='Product'
            style={style}
        />
    )
}