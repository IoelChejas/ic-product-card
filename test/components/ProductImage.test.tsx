import React from 'react'
import renderer from 'react-test-renderer'
import { ProductImage, ProductCard } from '../../src/components';
import { product1 } from '../data/products'

describe('ProductImage', () => {

    test('should show component with image', () => {

        const wrapper = renderer.create(
            <ProductImage img="https://hola.jpg" />
        )

        expect(wrapper.toJSON()).toMatchSnapshot();

    });

    test('should show component with product image', () => {

        const wrapper = renderer.create(
            <ProductCard product={product1} >
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();

    });

});
