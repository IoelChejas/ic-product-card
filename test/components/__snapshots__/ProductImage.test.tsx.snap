// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should show component with image 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;

exports[`ProductImage should show component with product image 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="test-file-stub"
  />
</div>
`;
